Cabal/Distribution/Parsec/Types/FieldParser.hs

```diff
@@ -0,0 +1,340 @@
+    :: (Parsec a, Text a)
+    => FieldNamee  -- ^ field name
+    -> (s -> a)   -- ^ getter for showing
+    -> FieldFancy s a
+uniqueField = error "define me"
+
+-- | Field which can be defined at most once.
+affineField
+    :: (Parsec a, Text a)
+    => FieldNamee
+    -> (s -> Maybe a)
+    -> FieldFancy s (Maybe a)
+affineField fn = affineFieldAla fn Identity
+
+affineFieldAla
+    :: forall a b s. (Parsec b, Text b, Newtype b a)
+    => FieldNamee
+    -> (a -> b) -- ^ 'pack'
+    -> (s -> Maybe a)
+    -> FieldFancy s (Maybe a)
+affineFieldAla fn _pack _extract = FieldFancy (Set.singleton fn) parser
+  where
+    parser :: Fields Position -> ParseResult (Maybe a)
+    parser fields = case Map.lookup fn fields of
+        Nothing -> pure Nothing
+        Just [] -> pure Nothing
+        Just [MkNamelessField pos fls] -> Just . (unpack :: b -> a) <$> runFieldParser pos parsec fls
+
+
+
+-- | Field which can be define multiple times, and the results are @mappend@ed.
+monoidField
+    :: (Parsec a, Text a, Monoid a)
+    => FieldNamee  -- ^ field name
+    -> (s -> a)   -- ^ getter for showing
+    -> FieldFancy s a
+monoidField = error "define me"
+
+monoidFieldAla
+    :: (Parsec b, Text b, Monoid a, Newtype b a)
+    => FieldNamee
+    -> (a -> b) -- ^ 'pack'
+    -> (s -> a) -- ^ getter for showing
+    -> FieldFancy s a
+monoidFieldAla = error "define me"
+
+-- | Parser matching all fields with a name starting with a prefix.
+prefixFieldFancy
+    :: FieldNamee -- ^Prefix
+    -> (s -> [(String, String)]) -- ^ getter for showing
+    -> FieldFancy s [(String, String)]
+prefixFieldFancy = error "define me"
+
+-------------------------------------------------------------------------------
+-- Newtype
+-------------------------------------------------------------------------------
+
+class Newtype n o | n -> o where
+    pack :: o -> n
+    unpack :: n -> o
+
+-- 
+newtype CommaListWithSep a = CommaListWithSep { getCommaListWithSep :: [a] }
+
+instance Newtype (CommaListWithSep a) [a] where
+    pack = CommaListWithSep
+    unpack = getCommaListWithSep
+
+instance Parsec a => Parsec (CommaListWithSep a) where
+    parsec = pack <$> parsecOptCommaList parsec
+
+instance Text a => Text (CommaListWithSep a) where
+
+--
+newtype Token = Token { getToken :: String }
+
+instance Parsec Token where
+    parsec = pack <$> parsecToken
+
+instance Newtype Token String where
+    pack = Token
+    unpack = getToken
+
+instance Text Token
+
+--
+newtype FreeText = FreeText { getFreeText :: String }
+
+instance Parsec FreeText where
+    parsec = pack <$> many Parsec.anyChar 
+
+instance Newtype FreeText String where
+    pack = FreeText
+    unpack = getFreeText
+
+instance Text FreeText
+
+--
+newtype FilePathNT = FilePathNT { getFilePathNT ::  FilePath }
+
+instance Parsec FilePathNT where
+    parsec = pack <$> parsecFilePath
+
+instance Newtype FilePathNT FilePath where
+    pack = FilePathNT
+    unpack = getFilePathNT
+
+instance Text FilePathNT
```

Comment:
Text instances will be replaced with `Display` class (I once tried to add, but I have to try again!)